src/tensors/wrapperize.jl

```diff
@@ -0,0 +1,187 @@
+
+struct DepthCalculator
+    rec
+end
+
+depth_calculator(root) = DepthCalculator(DepthCalculatorVisitor()(root))
+
+(ctx::DepthCalculator)(node::FinchNode) = maximum(node -> get(ctx.rec, node, 0), PostOrderDFS(node))
+
+@kwdef struct DepthCalculatorVisitor
+    depth = 1
+    rec = Dict()
+end
+
+function (ctx::DepthCalculatorVisitor)(node::FinchNode)
+    if node.kind === loop
+        ctx(node.ext)
+        ctx.rec[node.idx] = ctx.depth
+        ctx_2 = DepthCalculatorVisitor(depth=ctx.depth+1, rec=ctx.rec)
+        ctx_2(node.body)
+    elseif node.kind === define
+        ctx.rec[node.lhs] = ctx.depth
+    elseif istree(node)
+        for child in children(node)
+            ctx(child)
+        end
+    end
+    return ctx.rec
+end
+
+struct VariableExpander
+    bindings
+end
+
+function variable_expander(root)
+    bindings = Dict()
+    for node in PostOrderDFS(root)
+        if node.kind === define
+            bindings[node.lhs] = node.rhs
+        end
+    end
+    return VariableExpander(bindings)
+end
+
+(ctx::VariableExpander)(node::FinchNode) = Rewrite(Fixpoint(@rule ~var::isvariable => bindings[var]))(node)
+"""
+    get_wrapper_rules(alg, shash)
+
+Return the wrapperizing rule set for Finch, which converts expressions like `A[i
++ 1]` to array combinator expressions like `OffsetArray(A, (1,))`. The rules have
+access to the algebra `alg` and the depth lookup `depth`` One can dispatch on
+the `alg` trait to specialize the rule set for different algebras. These rules run
+after simplification so one can expect constants to be folded.
+"""
+function get_wrapper_rules(alg, depth, ctx)
```

Comment:
Is there any reason to be concerned about rewrite rule ordering or anything of the like here? I suppose if it causes problems one could just use the wrappers manually...